pkg/observability/k8s-events.go

```diff
@@ -0,0 +1,178 @@
+	spotITNReason                 = "SpotInterruption"
+	sqsTerminateReason            = "SQSTermination"
+	rebalanceRecommendationReason = "RebalanceRecommendation"
+	unknownReason                 = "UnknownInterruption"
+)
+
+// K8sEventRecorder wraps a Kubernetes event recorder with some extra information
+type K8sEventRecorder struct {
+	annotations map[string]string
+	enabled     bool
+	node        *corev1.Node
+	record.EventRecorder
+}
+
+// InitK8sEventRecorder creates a Kubernetes event recorder
+func InitK8sEventRecorder(enabled bool, nodeName string, nodeMetadata ec2metadata.NodeMetadata, extraAnnotationsStr string) (K8sEventRecorder, error) {
+	if !enabled {
+		return K8sEventRecorder{}, nil
+	}
+
+	// Create default annotations
+	// Worth iterating over nodeMetadata fields using reflect? (trutx)
+	annotations := make(map[string]string)
+	annotations["account-id"] = nodeMetadata.AccountId
+	annotations["availability-zone"] = nodeMetadata.AvailabilityZone
+	annotations["instance-id"] = nodeMetadata.InstanceID
+	annotations["instance-type"] = nodeMetadata.InstanceType
+	annotations["local-hostname"] = nodeMetadata.LocalHostname
+	annotations["local-ipv4"] = nodeMetadata.LocalIP
+	annotations["public-hostname"] = nodeMetadata.PublicHostname
+	annotations["public-ipv4"] = nodeMetadata.PublicIP
+	annotations["region"] = nodeMetadata.Region
+
+	// Parse extra annotations
+	var err error
+	if extraAnnotationsStr != "" {
+		annotations, err = parseExtraAnnotations(annotations, extraAnnotationsStr)
+		if err != nil {
+			return K8sEventRecorder{}, err
+		}
+	}
+
+	// Get in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return K8sEventRecorder{}, err
+	}
+
+	// Create clientSet
+	clientSet, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return K8sEventRecorder{}, err
+	}
+
+	// Get node
+	node, err := clientSet.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if err != nil {
+		return K8sEventRecorder{}, err
+	}
+
+	// Create broadcaster
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: clientSet.CoreV1().Events("default")})
+
+	// Create event recorder
+	return K8sEventRecorder{
+		annotations: annotations,
+		enabled:     true,
+		node:        node,
+		EventRecorder: broadcaster.NewRecorder(
+			scheme.Scheme,
+			corev1.EventSource{
+				Component: "aws-node-termination-handler",
+				Host:      nodeName,
+			},
+		),
+	}, nil
+}
+
+// Emit a Kubernetes event for the current node and with the given event type, reason and message
+func (r K8sEventRecorder) Emit(eventType, eventReason, eventMsgFmt string, eventMsgArgs ...interface{}) {
+	if r.enabled {
+		r.AnnotatedEventf(r.node, r.annotations, eventType, eventReason, eventMsgFmt, eventMsgArgs...)
+	}
+}
+
+// GetReasonForKind returns a Kubernetes event reason for the given interruption event kind
+func GetReasonForKind(kind string) string {
+	switch kind {
```

Comment:
should events from ASG lifecycle hooks be included here as well? such as [Scale-In](https://docs.aws.amazon.com/autoscaling/ec2/userguide/AutoScalingGroupLifecycle.html#as-lifecycle-scale-in)

Reply:
Not sure I follow. AFAICT all SQS-based events are stamped with `SQSTerminateKind` (i.e. here https://github.com/aws/aws-node-termination-handler/blob/main/pkg/monitor/sqsevent/asg-lifecycle-event.go#L74).

Haven't taken a deep look but if we wanted to be more fine-grained around SQS and/or ASG lifecycle events shouldn't we have to code additional logic in each SQS-based monitor to parse the actual reason coming from AWS and stamp it into the `interruptionEvent.Kind` field? If we did that we'd automatically have what you mention if I got things right. Can't help much on the SQS side because I have always used ANTH with IMDS.

However, if I did get things right I think such changes are outside the scope of this PR.

Reply:
Ah, thanks for the code pointer and explanations. Agreed the additional logic to do that is outside the scope of this PR-- resolving